src/MissionManager.cpp

```diff
@@ -106,17 +106,37 @@ void MissionManager::taskEnd(CNSocket* sock, CNPacketData* data) {
     // failed timed missions give an iNPC_ID of 0
     if (missionData->iNPC_ID == 0) {
         TaskData* task = MissionManager::Tasks[missionData->iTaskNum];
-        // double-checking
-        if (task->task["m_iSTGrantTimer"] > 0) {
+        if (task->task["m_iSTGrantTimer"] > 0) { // its a timed mission
             Player* plr = PlayerManager::getPlayer(sock);
-            int failTaskID = task->task["m_iFOutgoingTask"];
-            if (failTaskID != 0) {
-                MissionManager::quitTask(sock, missionData->iTaskNum, false);
+            /*
+             * Enemy killing missions
+             * this is gross and should be cleaned up later
+             * once we comb over mission logic more throughly
+             */
+            bool mobsAreKilled = false;
+            if (task->task["m_iHTaskType"] == 5) {
+                mobsAreKilled = true;
+                for (int i = 0; i < ACTIVE_MISSION_COUNT; i++) {
+                    if (plr->tasks[i] == missionData->iTaskNum) {
+                        for (int j = 0; j < 3; j++)
+                            if (plr->RemainingNPCCount[i][j] > 0)
+                                mobsAreKilled = false;
+                        break;
```

Comment:
This `break` statement is going to break out of the outer loop, that iterates through the missions, and **not** the inner one. I don't have enough context to know if that was what you intended, but just thought I'd point it out, since we usually avoid single-line control statements.

Reply:
I intended the outer loop but thinking about it, the inner loop would also be sensible.